File: common/src/main/java/com/ultreon/mods/advanceddebug/api/common/Percentage.java
package com.ultreon.mods.advanceddebug.api.common;

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext;

import java.util.Objects;

/**
 * Percentage class, used for get percentage value or normalized value.<br>
 * Use {@link #getPercentage() getPercentage} or {@link #setPercentage(double) setPercentage} for getting / setting the percentage value.<br>
 * Use {@link #getValue() getValue} or {@link #setValue(double) setValue} for getting / settings the normalized value.<br>
 * <br>
 *
 * @author XyperCode
 */
public final class Percentage implements IFormattable {
    private double percentage;

    public Percentage(double value) {
        this.percentage = value * 100;
    }

    public Percentage(int percentage) {
        this.percentage = percentage;
    }

    @Override
    public void format(IFormatterContext ctx) {
        ctx.number(Math.round(percentage)).operator("%");
    }

    public double getValue() {
        return percentage / 100;
    }

    public void setPercentage(double percentage) {
        this.percentage = percentage;
    }

    public void setValue(double value) {
        this.percentage = value * 100;
    }

    public Multiplier toMultiplier() {
        return new Multiplier(this.percentage / 100);
    }

    public double getPercentage() {
        return percentage;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Percentage that = (Percentage) o;
        return Double.compare(that.percentage, percentage) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(percentage);
    }

    @Override
    public String toString() {
        return "Percentage{" +
                "percentage=" + percentage +
                '}';
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/api/common/Percentage.kt
package com.ultreon.mods.advanceddebug.api.common

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext
import java.util.*
import kotlin.math.roundToLong

/**
 * Percentage class, used for get percentage value or normalized value.<br></br>
 * Use [getPercentage][.getPercentage] or [setPercentage][.setPercentage] for getting / setting the percentage value.<br></br>
 * Use [getValue][.getValue] or [setValue][.setValue] for getting / settings the normalized value.<br></br>
 * <br></br>
 *
 * @author Qboi123
 */
@Suppress("MemberVisibilityCanBePrivate")
class Percentage : IFormattable {
    var percentage: Double

    constructor(percentage: Double) {
        this.percentage = percentage
    }

    constructor(percentage: Int) {
        this.percentage = percentage.toDouble()
    }

    override fun format(ctx: IFormatterContext) {
        ctx.number(percentage.roundToLong()).operator("%")
    }

    var value: Double
        get() = percentage / 100
        set(value) {
            percentage = value * 100
        }

    val asMultiplier: Multiplier get() = (percentage / 100).multiplier

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as Percentage
        return that.percentage.compareTo(percentage) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(percentage)
    }

    override fun toString(): String {
        return "Percentage{" +
                "percentage=" + percentage +
                '}'
    }
}

val Double.asPercentage get() = (this * 100).percent
val Number.asPercentage get() = (this.toDouble() * 100).percent
val Double.percent get() = Percentage(this)
val Number.percent get() = this.toDouble().percent
